File: php/p797-allPathsSourceTarget.php
<?php

class Solution {

    /**
     * @param Integer[][] $graph
     * @return Integer[][]
     */
    function allPathsSourceTarget($graph) {

        $end = count($graph) - 1;
        $path = [];
        $seen = array_fill(0, $end+1, false);
        
        $stack = [];
        foreach ($graph[0] as $v) $stack[] = [0, $v];

        while ($stack) {

            //print_r($stack);
            $edge = array_pop($stack);

            $ev = end($edge);
            $seen[$ev] = true;

            if ($ev == $end) {
                $path[] = $edge;
                $seen[$ev] = false;
                continue;
            }

            foreach ($graph[$ev] as $nv) {
                if (!$seen[$nv]) $stack[] = array_merge($edge, [$nv]);
            }

            $seen[$ev] = false;
        }

        return $path;
    }
}

$graph = [[1,2],[3],[3],[]];
$graph = [[4,3,1],[3,2,4],[3],[4],[]];
$graph = [[1,2,3],[2],[3],[]];

$so = new Solution();
$r = $so->allPathsSourceTarget($graph);
print_r($r);

File: python/p797-allPathsSourceTarget.py
import pprint
pp = pprint.PrettyPrinter()
from var_dump import var_dump

class Solution:
    def allPathsSourceTarget(self, graph: list[list[int]]) -> list[list[int]]:
        
        def dfs(node: int):
            if node == n-1:
                r.append(path.copy())
                return

            for i in graph[node]:
                path.append(i)
                dfs(i)
                path.pop()


        n = len(graph)
        r = []
        path = [0]
        dfs(0)
        return r


graph = [[1,2],[3],[3],[]]
graph = [[4,3,1],[3,2,4],[3],[4],[]]

so = Solution()
r = so.allPathsSourceTarget(graph)
pp.pprint(r)